src/main.rs

```diff
@@ -869,4 +913,185 @@ mod tests {
+        const MILAN_CSR: &[u8] = include_bytes!("../crates/snp_validation/src/milan.signed.csr");
+
+        fn assert_sgx_config(
+            csr: &CertReq<'_>,
+            conf: &sgx_validation::config::Config,
+        ) -> anyhow::Result<()> {
+            let sgx = sgx_validation::Sgx::default();
+
+            #[allow(unused_variables)]
+            for Attribute { oid, values } in csr.info.attributes.iter() {
+                for any in values.iter() {
+                    let ereq: ExtensionReq<'_> = any.decode_into()?;
+                    for ext in Vec::from(ereq) {
+                        let (quote, bytes): (Quote<'_>, _) = ext.extn_value.parse()?;
+                        let chain = quote.chain()?;
+                        let chain = chain
+                            .iter()
+                            .map(|c| Certificate::from_der(c))
+                            .collect::<Result<Vec<_>, _>>()?;
+
+                        // Validate the report.
+                        let pck = sgx.trusted(&chain)?;
+                        let report = quote.verify(pck)?;
+                        sgx.verify(&csr.info, &ext, Some(conf), false)?;
+                    }
+                }
+            }
+            Ok(())
+        }
+
+        fn assert_snp_config(
+            csr: &CertReq<'_>,
+            conf: &snp_validation::config::Config,
+        ) -> anyhow::Result<()> {
+            let snp = Snp::default();
+            #[allow(unused_variables)]
+            for Attribute { oid, values } in csr.info.attributes.iter() {
+                for any in values.iter() {
+                    let ereq: ExtensionReq<'_> = any.decode_into()?;
+                    for ext in Vec::from(ereq) {
+                        let evidence = Evidence::from_der(ext.extn_value)?;
+                        let array = evidence.report.try_into()?;
+                        let report = Report::cast(array);
+                        snp.verify(&csr.info, &ext, Some(conf), false)?;
+                    }
+                }
+            }
+            Ok(())
+        }
+
+        #[test]
+        fn test_config_empty() {
+            let config: Config = toml::from_str("").expect("Couldn't deserialize");
+
+            assert!(config.sgx.is_none());
+            assert!(config.snp.is_none());
+        }
+
+        #[test]
+        fn test_config() {
+            let config: Config = toml::from_str(
+                r#"
+            [snp]
+            policy_flags = ["SingleSocket", "Debug"]
+            enarx_signer = ["1234567890", "00112233445566778899"]
+            minimum_abi = "1.0"
+            
+            [sgx]
+            enarx_signer = ["1234567890", "00112233445566778899"]
+            "#,
+            )
+            .expect("Couldn't deserialize");
+
+            assert!(config.sgx.is_some());
+            assert!(config.snp.is_some());
+            assert_eq!(config.sgx.unwrap().measurements.signer.len(), 2);
```

Comment:
This doesn't really test much, you should assert whatever fields you set in the config, otherwise why set them?

Reply:
Just to test deserializing `snp` and `sgx` in one file, the other tests the fields of those structs.

Reply:
But how do you know that the `snp` config is still valid when you have both specified?
Maybe your deserializer completely discards the one or other config when both technologies are set or whatever other strange bugs.

This is essentially an integration test and that's the most important kind of test. Tests in other subcrates don't really matter - those are just unit tests used to facilitate development and test some very specific behaviors.
Integration tests, however, are the tests we really care about, because they test the system as a whole and are the closest to the behavior that this service will exhibit in production setting. They really should be complete.

Reply:
Done, now all fields are compared to be sure parsing worked as expected. Please check.